nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,341 @@
+"""
+region_signal_extractor
+"""
+
+import nibabel
+import numpy as np
+
+from scipy.ndimage import label
+from scipy.stats import scoreatpercentile
+
+from sklearn.base import clone
+from sklearn.externals.joblib import Memory
+
+from skimage.feature import peak_local_max
+from skimage.segmentation import random_walker
+
+from nilearn import plotting
+from nilearn.input_data import NiftiMasker, NiftiMapsMasker
+from nilearn._utils import check_niimg, check_niimg_3d, check_niimg_4d
+from nilearn._utils.extmath import fast_abs_percentile
+from nilearn.image import iter_img, new_img_like
+from nilearn.image.image import _smooth_array
+from nilearn._utils.niimg_conversions import concat_niimgs
+from nilearn._utils.compat import _basestring
+
+
+def extract_regions(map_data, affine, label_data, min_size):
+    """ Extract each region from the connected components
+        data where the data is already assigned with a
+        unique label point.
+
+    Parameters
+    ----------
+    label_data: a numpy array
+        an array of data where each data point is labelled.
+
+    Returns
+    -------
+    regions_img: a 4D Nifti like image/object
+        contains the regions extracted from
+        the labelled data. Each extracted region
+        lies in a list of 3D volumes.
+    """
+    regions = []
+
+    labels_size = np.bincount(label_data.ravel())
+    labels_size[0] = 0.
+
+    for label_id, label_size in enumerate(labels_size):
+        if label_size > min_size:
+            region = (label_data == label_id) * map_data
+            region_img = new_img_like(nibabel.Nifti1Image(map_data, affine),
```

Comment:
More explanation : the "img" object should be created at the very end of the pipeline. Also, I like to separate "nifti world" and "data world". Here, your function takes data as input and output niftis. Plus, you use `new_img_like` with a freshly new image, which makes the call to this function completely counterproductive.
You should just do `regions.append(map_data)` and turn the data into niftis in the upper level function.